src/Datagrid/Pager.php

```diff
@@ -143,64 +157,36 @@ public function init()
     /**
      * NEXT_MAJOR: remove this method.
      *
-     * @deprecated since sonata-project/doctrine-orm-admin-bundle 3.27
-     *
      * @param int $nb
+     *
+     * @deprecated since sonata-project/doctrine-orm-admin-bundle 3.27
      */
     protected function setNbResults($nb)
     {
         if ('sonata_deprecation_mute' !== (\func_get_args()[1] ?? null)) {
-            @trigger_error(sprintf(
-                'The %s() method is deprecated since sonata-project/doctrine-orm-admin-bundle 3.27 and will be removed in 4.0. Use "setResultsCount()" instead.',
-                __METHOD__,
-            ), E_USER_DEPRECATED);
+            @trigger_error(
+                sprintf(
+                    'The %s() method is deprecated since sonata-project/doctrine-orm-admin-bundle 3.27 and will be removed in 4.0. Use "setResultsCount()" instead.',
+                    __METHOD__,
+                ),
+                E_USER_DEPRECATED
+            );
         }
 
         $this->nbResults = $nb;
         $this->resultsCount = (int) $nb;
     }
 
-    private function countCompositePrimaryKey(ProxyQueryInterface $countQuery): void
-    {
-        $rootAliases = current($countQuery->getRootAliases());
-        $countQuery->setParameter('concat_separator', self::CONCAT_SEPARATOR);
-
-        $columns = $rootAliases.'.'.implode(', :concat_separator, '.$rootAliases.'.', $this->getCountColumn());
-
-        $countQuery->select(sprintf(
-            'count(%s concat(%s)) as cnt',
-            $countQuery instanceof ProxyQuery && !$countQuery->isDistinct() ? null : 'DISTINCT',
-            $columns
-        ));
-    }
-
-    private function countSinglePrimaryKey(ProxyQueryInterface $countQuery): void
-    {
-        $countQuery->select(sprintf(
-            'count(%s %s.%s) as cnt',
-            $countQuery instanceof ProxyQuery && !$countQuery->isDistinct() ? null : 'DISTINCT',
-            current($countQuery->getRootAliases()),
-            current($this->getCountColumn())
-        ));
-    }
-
     private function computeResultsCount(): int
     {
-        $countQuery = clone $this->getQuery();
-
-        if (\count($this->getParameters()) > 0) {
-            $countQuery->setParameters($this->getParameters());
+        if (!$this->getQuery() instanceof ProxyQueryInterface) {
```

Comment:
What's the purpose of this check ?

`getQuery` is returning `ProxyQueryInterface|null`.
It seems better to write `null !== $this->getQuery()`.
And then, the message should explain that the query wasn't set.

If the goal is to check that the method `getQueryBuilder` exists, 
you should not use `Sonata\AdminBundle\Datagrid\ProxyQueryInterface` but `Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQueryInterface`

Reply:
> But then, maybe we should directly add these checks in `setQuery/getQuery`.

Maybe Pager should be generic 
```
/**
 * @template Query of ProxyQueryInterface
 */
```

WDYT @sonata-project/contributors ?

Reply:
I created https://github.com/sonata-project/SonataAdminBundle/pull/6759